### ISSUE TITLE
lancero driver failes to compile in ubuntu20

### ISSUE THREAD

**Issue by joefowler** [usnistgov/quantum-sensors-drivers#13]
**[Original report](https://bitbucket.org/nist_microcal/nist_qsp_drivers/issues/13) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

```
pcuser@687horton:~/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver$ make
make -C /lib/modules/5.11.0-38-generic/build M=/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver modules
make[1]: Entering directory '/usr/src/linux-headers-5.11.0-38-generic'
  CC [M]  /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-base.o
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-base.c: In function ‘char_events_read’:
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-base.c:2853:2: warning: ignoring return value of ‘copy_to_user’, declared with attribute warn_unused_result [-Wunused-result]
 2853 |  copy_to_user(buf, &lro->events_irq, 4);
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-base.c: In function ‘char_ctrl_write’:
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-base.c:2818:2: warning: ignoring return value of ‘copy_from_user’, declared with attribute warn_unused_result [-Wunused-result]
 2818 |  copy_from_user(&w, buf, 4);
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-base.c: In function ‘char_ctrl_read’:
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-base.c:2798:2: warning: ignoring return value of ‘copy_to_user’, declared with attribute warn_unused_result [-Wunused-result]
 2798 |  copy_to_user(buf, &w, 4);
      |  ^~~~~~~~~~~~~~~~~~~~~~~~
  CC [M]  /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.o
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c: In function ‘sgm_get_user_pages’:
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:147:26: error: ‘struct mm_struct’ has no member named ‘mmap_sem’; did you mean ‘mmap_base’?
  147 |  down_read(&current->mm->mmap_sem);
      |                          ^~~~~~~~
      |                          mmap_base
/home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.c:159:24: error: ‘struct mm_struct’ has no member named ‘mmap_sem’; did you mean ‘mmap_base’?
  159 |  up_read(&current->mm->mmap_sem);
      |                        ^~~~~~~~
      |                        mmap_base
make[2]: *** [scripts/Makefile.build:288: /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver/lancero-user.o] Error 1
make[1]: *** [Makefile:1849: /home/pcuser/nist_qsp_drivers/lancero/Lancero-RELEASE/C/driver] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.11.0-38-generic'
make: *** [Makefile:32: lancero.ko] Error 2
```

‌

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Based on [https://www.spinics.net/lists/linux-mm/msg214191.html](https://www.spinics.net/lists/linux-mm/msg214191.html)

‌

I found that the following patch compiles on ubuntu 20. For now I’ll just commit it to a branch, but probably we should add some macro thing to check the kernel version?

```none
diff --git a/lancero/Lancero-RELEASE/C/driver/lancero-user.c b/lancero/Lancero-RELEASE/C/driver/lancero-user.c
index 4320f36..211f03b 100644
--- a/lancero/Lancero-RELEASE/C/driver/lancero-user.c
+++ b/lancero/Lancero-RELEASE/C/driver/lancero-user.c
@@ -144,7 +144,8 @@ int sgm_get_user_pages(struct sg_mapping_t *sgm, const char *start, size_t count
                pages[i] = NULL;
        }
        /* try to fault in all of the necessary pages */
-       down_read(&current->mm->mmap_sem);
+       //down_read(&current->mm->mmap_base);
+       mmap_read_lock(current->mm);
        /* to_user != 0 means read from device, write into user space buffer memory */
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4,6,0)   // AIO API is changed in Linux 4.6 (JWF 9 May 2018)
@@ -156,7 +157,8 @@ int sgm_get_user_pages(struct sg_mapping_t *sgm, const char *start, size_t count
 #else
        rc = get_user_pages((unsigned long)start, nr_pages, gup_flags, pages, NULL);
 #endif
-       up_read(&current->mm->mmap_sem);
+       //up_read(&current->mm->mmap_base); // works on kernel versions 4?
+       mmap_read_unlock(current->mm);
 
        /* errors and no page mapped should return here */
        if (rc < nr_pages) {
```

‌

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

What exact version of the kernel needs these changes? \(That’s always the killer question!\)

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Test on Linux kernel version to fix problems w/ v5.8+

Fixes #13
